src/state.rs

```diff
@@ -308,88 +163,54 @@ pub struct Allowance {
 impl Allowance {
     pub fn is_expired_at(&self, block: &cosmwasm_std::BlockInfo) -> bool {
         match self.expiration {
-            Some(time) => block.time >= time,
+            Some(time) => block.time.seconds() >= time,
             None => false, // allowance has no expiration
         }
     }
 }
 
-pub fn read_allowance<S: Storage>(
-    store: &S,
-    owner: &CanonicalAddr,
-    spender: &CanonicalAddr,
-) -> StdResult<Allowance> {
-    let owner_store =
-        ReadonlyPrefixedStorage::multilevel(&[PREFIX_ALLOWANCES, owner.as_slice()], store);
-    let owner_store = TypedStore::attach(&owner_store);
-    let allowance = owner_store.may_load(spender.as_slice());
-    allowance.map(Option::unwrap_or_default)
-}
+pub static ALLOWANCES: Keymap<(Addr, Addr), Allowance> = Keymap::new(PREFIX_ALLOWANCES);
+pub struct AllowancesStore {}
+impl AllowancesStore {
+    pub fn may_load(store: &dyn Storage, owner: &Addr, spender: &Addr) -> StdResult<Allowance> {
+        ALLOWANCES
+            .get(store, &(owner.clone(), spender.clone()))
+            .ok_or_else(|| StdError::generic_err(""))
+    }
 
-pub fn write_allowance<S: Storage>(
-    store: &mut S,
-    owner: &CanonicalAddr,
-    spender: &CanonicalAddr,
-    allowance: Allowance,
-) -> StdResult<()> {
-    let mut owner_store =
-        PrefixedStorage::multilevel(&[PREFIX_ALLOWANCES, owner.as_slice()], store);
-    let mut owner_store = TypedStoreMut::attach(&mut owner_store);
-
-    owner_store.store(spender.as_slice(), &allowance)
+    pub fn save(
+        store: &mut dyn Storage,
+        owner: &Addr,
+        spender: &Addr,
+        allowance: &Allowance,
+    ) -> StdResult<()> {
+        ALLOWANCES.insert(store, &(owner.clone(), spender.clone()), allowance)
+    }
 }
 
 // Viewing Keys
 
-pub fn write_viewing_key<S: Storage>(store: &mut S, owner: &CanonicalAddr, key: &ViewingKey) {
-    let mut balance_store = PrefixedStorage::new(PREFIX_VIEW_KEY, store);
+pub fn write_viewing_key(store: &mut dyn Storage, owner: &CanonicalAddr, key: &ViewingKey) {
+    let mut balance_store = PrefixedStorage::new(store, PREFIX_VIEW_KEY);
     balance_store.set(owner.as_slice(), &key.to_hashed());
 }
 
-pub fn read_viewing_key<S: Storage>(store: &S, owner: &CanonicalAddr) -> Option<Vec<u8>> {
-    let balance_store = ReadonlyPrefixedStorage::new(PREFIX_VIEW_KEY, store);
+pub fn read_viewing_key(store: &dyn Storage, owner: &CanonicalAddr) -> Option<Vec<u8>> {
+    let balance_store = ReadonlyPrefixedStorage::new(store, PREFIX_VIEW_KEY);
     balance_store.get(owner.as_slice())
 }
```

Comment:
You can use the toolkit's `ViewingKey` interface, it's already implemented there.
Keep that in mind when you port `viewing_key.rs` as well (I think it's mostly redundant).